Fixes #17443: Adds ExportTemplate.file_name field

jnovinger opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note that
    our contribution policy requires that a feature request or bug report be
    approved and assigned prior to opening a pull request. This helps avoid
    waste time and effort on a proposed change that we might not be able to
    accept.

    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ISSUE WHICH HAS BEEN ASSIGNED
    TO YOU, IT WILL BE CLOSED AUTOMATICALLY.

    Please specify your assigned issue number on the line below.
-->
### Fixes: #17443 

- adds `ExportTemplate.file_name` field
- adds logic to `ExportTemplate.render_to_response()` to use `file_name` if populated
- adds field to required places (API serializers, forms, filterset, etc.)
- updates tests for new field
- adds a new utils, `filename_from_model`, to standardize generating export filenames for models
- adds tests for new util func
<!--
    Please include a summary of the proposed changes below.
-->

jnovinger committed:
Fixes #17443: Adds ExportTemplate.file_name field

jeremystretch reviewed (changes requested):
Thanks @jnovinger! Just had some very minor cleanup suggestions.

jnovinger commented on netbox/extras/models/models.py:
```diff
@@ -415,14 +415,19 @@ class ExportTemplate(SyncedDataMixin, CloningMixin, ExportTemplatesMixin, Change
         blank=True,
         help_text=_('Extension to append to the rendered filename')
     )
+    file_name = models.CharField(
```
Yep! Not sure why I did elsewhere but not here. Derp.

jnovinger commented on netbox/extras/models/models.py:
```diff
@@ -415,14 +415,19 @@ class ExportTemplate(SyncedDataMixin, CloningMixin, ExportTemplatesMixin, Change
         help_text=_('Extension to append to the rendered filename')
     )
+    file_name = models.CharField(
+        max_length=1000,
```
Thanks, this was a guess based on what I saw elsewhere. But, I think that other example was probably for a full path as opposed to just a filename.

jnovinger commented on netbox/extras/forms/bulk_edit.py:
```diff
@@ -155,6 +155,10 @@ class ExportTemplateBulkEditForm(BulkEditForm):
         max_length=50,
         required=False
     )
+    file_name = forms.CharField(
```
Any guidance on when a field should or should not be included in `nullable_fields`?

jeremystretch replied:
Ideally any optional, non-M2M field on the model should be nullable, I think. There are probably some exceptions to that rule but none come to mind offhand.

jnovinger committed:
Addresses PR feedback

- Adds `file_name` to `ExportTemplateBulkEditForm.nullable_fields`
- Shortens max length of `ExportTemplate.file_name` to 200 chars
- Adds tests for `ExportTemplateFilterSet.file_extension`

jnovinger commented:
@jeremystretch , rather than adding another schema migration, I edited the existing migration to reflect the new max length for `ExportTemplate.file_name`. Apologies for making testing any more complicated. If you'd prefer it the other way, just let me know.

jnovinger commented:
Ah, looks like there are some conflicting migrations from a recent merge to `feature`. Fixing that up now.

jnovinger committed:
Fixes migration conflict caused by fix for #17841

jeremystretch reviewed (approved):
Thanks @jnovinger!
